# MACHINE LEARNING
## RANDOM FOREST REGRESSOR

Random Forest Regressor (RFR) is an ensemble learning algorithm that combines the prediction of multiple decision trees to improve the accuracy of the model.

Here's how RFR works:

1. **Random Sampling**: RFR starts by randomly selecting a subset of data from the training set.

2. __Decision Tree Creation__: Then, RFR creates a decision tree on this randomly selected data. The tree is created by recursively splitting the data into smaller subsets based on the features that best separate the data into target values.

3. **Tree Ensemble**: The RFR algorithm repeats the above two steps (random sampling and decision tree creation) to create a large number of decision trees.

4. **Prediction**: To make a prediction, RFR takes the average prediction of all decision trees. Each decision tree predicts a value for the target variable based on the features of the input data. The final prediction of the RFR is the average of all the predictions made by each decision tree.

5. **Bagging**: RFR also employs a technique called bagging to reduce overfitting. Bagging randomly selects subsets of features for each decision tree, which reduces the correlation between the trees and helps to improve the accuracy of the model.

Overall, RFR is a powerful algorithm that can handle both regression and classification problems. It is particularly effective when dealing with large datasets that have many features, and it can provide accurate predictions even in the presence of noise or missing data.

### Implementing in python

Random Forest Regressor is used to predict house prices of houses in Melbourne, Australia below.

In [3]:
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error

In [5]:
melbourne_data = pd.read_csv('House Prices\melb_data.csv')
melbourne_data.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  13580.000000  1.358000e+04  13580.000000  13580.000000  13580.000000   
mean       2.937997  1.075684e+06     10.137776   3105.301915      2.914728   
std        0.955748  6.393107e+05      5.868725     90.676964      0.965921   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.500000e+05      6.100000   3044.000000      2.000000   
50%        3.000000  9.030000e+05      9.200000   3084.000000      3.000000   
75%        3.000000  1.330000e+06     13.000000   3148.000000      3.000000   
max       10.000000  9.000000e+06     48.100000   3977.000000     20.000000   

           Bathroom           Car       Landsize  BuildingArea    YearBuilt  \
count  13580.000000  13518.000000   13580.000000   7130.000000  8205.000000   
mean       1.534242      1.610075     558.416127    151.967650  1964.684217   
std        0.691712      0.962634    3990.669241    541.014538    37.273762   
min        0.000000      0.000000       0.000000      0.000000  1196.000000   
25%        1.000000      1.000000     177.000000     93.000000  1940.000000   
50%        1.000000      2.000000     440.000000    126.000000  1970.000000   
75%        2.000000      2.000000     651.000000    174.000000  1999.000000   
max        8.000000     10.000000  433014.000000  44515.000000  2018.000000   

          Lattitude    Longtitude  Propertycount  
count  13580.000000  13580.000000   13580.000000  
mean     -37.809203    144.995216    7454.417378  
std        0.079260      0.103916    4378.581772  
min      -38.182550    144.431810     249.000000  
25%      -37.856822    144.929600    4380.000000  
50%      -37.802355    145.000100    6555.000000  
75%      -37.756400    145.058305   10331.000000  
max      -37.408530    145.526350   21650.000000

In [6]:
melbourne_data.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [8]:
#Splitting the data into X (features) and y (target variable)
X = melbourne_data[['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']].dropna(axis = 0)
y = melbourne_data.Price

In [17]:
#Splitting the data set into training and testing DataSets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.3, train_size = 0.7, random_state = 1)

In [25]:
#Creating an instance of the model 
model = RandomForestRegressor(random_state = 1)

In [26]:
#Fitting the model to the training model
model.fit(X_train, y_train)

RandomForestRegressor(random_state=1)

In [27]:
#Evaluating the model
score = model.score(X_test, y_test)
print(score)

0.7710555110047177


In [28]:
#Using the model to predict on unseen data
y_pred = model.predict(X_test)

In [29]:
#Calculating the mean absolute error 
print(mean_absolute_error(y_test, y_pred))

183910.24657987937


## Conclusion

An evaluation score of 0.771 and a mean absolute error of 183910.24 is a significant improvement over the previous evaluation score and mean absolute error of the [Decision Tree Regressor](https://github.com/gacheruh/ML-DecisionTreeRegressor/blob/main/House_Prices.ipynb). This suggests that the RandomForestRegressor model is better suited for the prediction of house prices than the Decision Tree Regressor.

The improvement in the evaluation score and mean absolute error can be attributed to the random forest algorithm's ability to reduce overfitting and variance in the model. It also performs feature selection, which helps to eliminate irrelevant or redundant features that could negatively impact the model's performance.

Overall, the RandomForestRegressor model seems to be a better choice for the task of predicting house prices compared to the Decision Tree Regressor.